test/functional/url-expander/test-expander.js

```diff
@@ -140,108 +140,248 @@ describes.realWin('Expander', {
-      const url = 'CONCAT(bad`hello`, world)';
-      allowConsoleError(() => { expect(() => {
-        expander.expand(url, mockBindings);
-      }).to.throw(/bad/); });
+    describe('async', () => {
+      it('should handle empty urls', () =>
+        expect(expander.expand('', mockBindings)).to.eventually.equal('')
+      );
+
+      it('parses one function, one argument', () =>
+        expect(expander.expand('TRIM(aaaaa    )', mockBindings))
+            .to.eventually.equal('aaaaa')
+      );
+
+      it('parses nested function one level', () =>
+        expect(expander.expand('UPPERCASE(TRIM(aaaaa    ))', mockBindings))
+            .to.eventually.equal('AAAAA')
+      );
+
+      it('parses nested function two levels', () =>
+        expect(expander.expand('LOWERCASE(UPPERCASE(TRIM(aAaA    )))',
+            mockBindings)).to.eventually.equal('aaaa')
+      );
+
+      it('parses one function, two string arguments', () =>
+        expect(expander.expand('CONCAT(aaa,bbb)', mockBindings))
+            .to.eventually.equal('aaabbb')
+      );
+
+      it('parses one function, two string arguments with space', () =>
+        expect(expander.expand('CONCAT(aaa , bbb)', mockBindings))
+            .to.eventually.equal('aaabbb')
+      );
+
+      it('parses function with func then string as args', () =>
+        expect(expander.expand('CONCAT(UPPERCASE(aaa),bbb)', mockBindings))
+            .to.eventually.equal('AAAbbb')
+      );
+
+      it('parses function with macro then string as args', () =>
+        expect(expander.expand('CONCAT(CANONICAL_URL,bbb)', mockBindings))
+            .to.eventually.equal('www.google.combbb')
+      );
+
+      it('parses function with string then func as args', () =>
+        expect(expander.expand('CONCAT(aaa, UPPERCASE(bbb))', mockBindings))
+            .to.eventually.equal('aaaBBB')
+      );
+
+      it('parses function with two funcs as args', () => {
+        const url = 'CONCAT(LOWERCASE(AAA),UPPERCASE(bbb)';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('aaaBBB');
+      });
+
+      it('parses function with three funcs as args', () => {
+        const url = 'CAT_THREE(LOWERCASE(AAA),UPPERCASE(bbb),LOWERCASE(CCC))';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal('aaaBBBccc');
+      });
+
+      it('should handle real urls', () => {
+        const url = 'http://www.amp.google.com/?client=CLIENT_ID(__ga)&canon=CANONICAL_URL&random=RANDOM';
+        const expected = 'http://www.amp.google.com/?client=amp-GA12345&canon=www.google.com&random=123456';
+        return expect(expander.expand(url, mockBindings))
+            .to.eventually.equal(expected);
+      });
+
+      it('should treat unrecognized keywords as normal strings', () => {
+        return expect(expander.expand('TRIM(FAKE(aaaaa))', mockBindings))
+            .to.eventually.equal('');
+      });
+
```

Comment:
can we have test case for wrong num of params
`CONCAT(abc)`

Reply:
what do you expect the result to be here? `fooundefined`? I think this should be up to the macro, not the `Expander`.

Reply:
talked offline. `fooundefined` it is. done.